tree/tree/src/TTree.cxx

```diff
@@ -4802,6 +4814,53 @@ Int_t TTree::FlushBaskets() const
    Int_t nerror = 0;
    TObjArray *lb = const_cast<TTree*>(this)->GetListOfBranches();
    Int_t nb = lb->GetEntriesFast();
+
+#ifdef R__USE_IMT
+   if (ROOT::IsImplicitMTEnabled() && fIMTEnabled) {
+      if (fSortedBranches.empty()) { const_cast<TTree*>(this)->InitializeSortedBranches(); }
+
+      BoolRAIIToggle sentry(fIMTFlush);
+      fIMTZipBytes.store(0);
+      fIMTTotBytes.store(0);
+      std::atomic<Int_t> nerrpar(0);
+      std::atomic<Int_t> nbpar(0);
+      std::atomic<Int_t> pos(0);
+      tbb::task_group g;
```

Comment:
We must be extremely careful when embedding calls to TBB into legacy code. Holding any shared resource (which a C++ object itself may be viewed as such a resource) can lead to deadlocks. See

https://software.intel.com/en-us/forums/intel-threading-building-blocks/topic/401006

for details. This is one reason why the CMS threaded-framework is built using explicit tasks rather than implicit ones. For the case of writing, I can't think of anyone being dependent on the TFile, although we must be certain that no ROOT locks are being held by the thread calling `g.wait()`.
The use of a `tbb::task_arena` would avoid the deadlock but does not appear to give us a way to limit the number of available threads.